doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+(This is why memory is included on the above charts).
+- In general, once we have read the entire snapshot from upstream, we process the remaining data quite quickly. This is particularly true for non-UPSERT sources.
+- Categorically fixing this means implementing _Hydration Backpressure_, which effectively locks Materialize's _read-rate_ to its _process-rate_.
+This is considered [out of scope](#out-of-scope) for this design, but, _critically_: if and when _Hydration backpressure_ is implemented, these proposed
+metrics _become even more useful_, as opposed to becoming irrelevant.
+
+
+## Steady-state
+
+The second set of metrics this design document proposes describe the _rate_ at which upstream and Materialize are processing data.
+These metrics are designed to answer #2 in [the problem statement](#the-problem).
+
+```
+| `upstream_rate`      | [`uint8`] | The rate at which data is appearing in the upstream service.       |
+| `process_rate`       | [`uint8`] | The rate at which Materialize is processing upstream data.         |
+```
+
+This rate has units that depend on the source type (replication bytes for Postgres, offsets for Kafka, gtid's for MySQL) and are
+designed to be _compared_. The following charts show what these metrics will look like, depending on whether Materialize is keeping up
+or falling behind.
+
+![steady-state](./static/source_metrics_2/steady_state.png)
+
+
+Note that when our _process rate_ is consistently below the `upstream_rate`, the user knows their source isn't keeping up with upstream.
+
+### Implementation
+
+These rates are _source frontier differences_. Each source's `SourceReader` implementation will
+be required to expose a continuous stream of _upstream source frontiers_. This means:
+
+- The Kafka source will periodically fetch metadata and expose per-partition offset high-watermarks.
+- The MySQL source will periodically expose the result of `SELECT @gtid_executed`, which is the gtid of the latest transaction.
+- The Postgres source will periodically expose the result of `SELECT pg_current_wal_lsn()`, which is the lsn of the latest transaction.
+
+Additionally, the source pipeline will periodically invert the latest frontier we have committed for the source, from the Materialize timestamp
+domain to the source-timestamp domain. For MySQL and Postgres sources, this frontier will the `meet` of the subsource frontiers.
+
+`upstream_rate` will be derived from the two latest _upstream source frontiers_, by calculating _source frontier difference_ between them.
+Similarly `process_rate` will be derived from the two latest committed frontiers.
+
+
+### Existing metrics
+
+Existing metrics such as `messages_received` and `bytes_received` continue to be useful for understanding the general performance of a given source.
+If users find some of them confusing or not useful, we can remove them in the future.
+
+## Minimal Viable Prototype
+
+N/A for now. This design document is primarily designed to expose
```

Comment:
Was there more you were going to say here?

Reply:
oh rip, i reviewed everything else!